client/src/Features/MicrosoftLearn/MicrosoftLearnFilterCore.ts

```diff
@@ -0,0 +1,199 @@
+    filteredProducts = new Set([...parents, ...products]);
+
+    let displayFilters: Map<FilterType, string[]> = new Map<FilterType, string[]>();
+    displayFilters.set(FilterType.Product, [...filteredProducts]);
+    displayFilters.set(FilterType.Role, [...filteredRoles]);
+    displayFilters.set(FilterType.Level, [...filteredLevels]);
+    displayFilters.set(FilterType.Type, [...filteredTypes]);
+
+    return displayFilters;
+}        
+
+export const updateURI = (selectedFilters: Map<FilterType, string[]>, searchTerm: string, expandedProducts: string[], productHierarchicalMap: Map<Product, Product[]>) => {
+    let productFilter = selectedFilters.get(FilterType.Product) || [];
+    let roleFilter = selectedFilters.get(FilterType.Role) || [];
+    let typeFilter = selectedFilters.get(FilterType.Type) || [];
+    let levelFilter = selectedFilters.get(FilterType.Level) || [];
+
+    const getProductUri = () => {
+      let parents = [...productHierarchicalMap.keys()]; 
+      let keep: string[] = [];
+      let parentProductFilters: Product[] = parents.filter(p => productFilter.includes(p.id));
+      let invisibleChildren = _.flatten(parentProductFilters.map(parent => productHierarchicalMap.get(parent)?.map(c=>c.id)));
+
+      keep = [...keep, ...productFilter.filter(f => !keep.includes(f) && !invisibleChildren.includes(f))];
+      return 'products='+keep.join('%2C');
+    }
+
+    let productUri = productFilter.length>0? getProductUri() : '';
+    let roleUri = roleFilter.length>0? 'roles=' + roleFilter.join('%2C') : '';
+    let typeUri = typeFilter.length>0? 'types=' + typeFilter.join('%2C') : '';
+    let levelUri = levelFilter.length>0? 'levels=' + levelFilter.join('%2C') : '';
+    let termsUri = searchTerm.length>0? 'terms='+searchTerm : '';
+    let expandedProductsUri = expandedProducts.length>0? 'expanded='+expandedProducts.join('%2C') : '';
+    let finalUri = [productUri, roleUri, levelUri, typeUri, termsUri, expandedProductsUri]
+                    .filter(s=>s.length!==0).join('&')            
+
+    let uri = window.location.protocol + "//" + window.location.host + window.location.pathname + '?' + finalUri;
+    window.history.pushState({path:uri},'',uri);       
+
+    return finalUri;
+}
+
+const getSearchTermFilteredLearnContent = (expressions: RegExp[], content: LearnContent[]): LearnContent[] => {
+    return content
+      .map(course => ({
+        course: course,
+        score: _.sumBy(
+          expressions,
+          singleExpression =>
+            scoreRegex(course.summary, singleExpression) + scoreRegex(course.title, singleExpression, 2)
+        )
+      }))
+      .filter(scouredCourse => scouredCourse.score > 0)
+      .sort((a, b) => b.score - a.score)
+      .map(scoredCourse => scoredCourse.course);
+}
+
+export const loadFiltersFromPath = (catalog: Catalog, productHierarchicalMap: Map<Product, Product[]>) => {
+    let searchParams = window.location.search
```

Comment:
My recommendation would be to do this from react's shadow DOM and not touch `window.*` properties. Per my understanding `ReactRouter` should be passing these values to you as `props | context` on your components. We should capture these values from there and pass them over to this function via store (I presume). Another downside of the current approach is that this function cannot be moved directly out of the browser window (such as a unit-test environment) at this moment, which essentially means points us to the fact that there's something wrong happening here.

> **Note:** We should try everything in the react counterparts of the things and refrain from manipulating the DOM directly.